File: Problem/p075_SortColors/Solution2.cpp
// Solution2: count and fill
#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
    void sortColors(vector<int>& nums) {
    	int zero = 0, one = 0;
    	
        for(int x : nums) {
			if(x == 0) zero++;	// count number of 0
			else if(x == 1) one++;	// count number of 1
		}
		fill(nums.begin(), nums.begin()+zero, 0);	// fill 0
		fill(nums.begin()+zero, nums.begin()+zero+one, 1);	// fill 1
		fill(nums.begin()+zero+one, nums.end(), 2);		// fill 2
		
//		this->show(nums);
		return;
    }
	
	void show(vector<int> nums) {
		for(int x : nums) {
			cout << x << " ";
		}
		cout << endl;
	}
};

int main() {
	
	Solution sol;
	vector<int> nums{1,0,2,2,0,1,1};
	
	sol.show(nums);
	
	sol.sortColors(nums);
	
	system("PAUSE");
	return 0;
} 

File: Problem/p091_DecodeWays/Solution2.java
package p091_DecodeWays;

public class Solution2 {
	public static void main(String[] args) {
		Solution2 sol = new Solution2();
		String s = "12";
		System.out.println(sol.numDecodings(s));
	}
	
	public int numDecodings(String s) {
        int len = s.length();
        int[] dp = new int[len+1];
        
        dp[len] = 1;
        for(int i = len-1; i >= 0; i--) {
        	if(s.charAt(i) != '0') {
        		dp[i] += dp[i+1];
        		if(i < len-1 && (s.charAt(i) == '1' || (s.charAt(i) == '2' && s.charAt(i+1) <= '6'))) {
        			dp[i] += dp[i+2];
        		}
        	}
        }
        
        show(dp);
        return dp[0];
    }
	
	// show int 1D-array
	private void show(int A[]) {
		int len = A.length;
		System.out.print("[");
		for(int i = 0; i < len; i++) {
			System.out.print(A[i]);
			if(i != len-1) {
				System.out.print(", ");
			}
		}
		System.out.print("]");
		System.out.println();
	}
}
